query/src/main/scala/filodb/query/exec/rangefn/LabelFunction.scala

```diff
@@ -0,0 +1,77 @@
+package filodb.query.exec.rangefn
+
+import java.util.regex.PatternSyntaxException
+
+import filodb.core.query.{CustomRangeVectorKeyWithShards, RangeVectorKey}
+import filodb.memory.format.ZeroCopyUTF8String
+import filodb.query.MiscellaneousFunctionId
+import filodb.query.MiscellaneousFunctionId.LabelReplace
+
+trait LabelTypeInstantFunction {
+
+  /**
+    * Apply the required instant function against the given value.
+    *
+    * @param value RangeVectorKey against which the function will be applied
+    * @return Updated RangeVectorKey
+    */
+  def apply(value: RangeVectorKey): RangeVectorKey
+
+}
+
+object LabelFunction {
+
+  /**
+    * This function returns a function that can be applied to generate the result.
+    *
+    * @param function to be invoked
+    * @param funcParams - Additional required function parameters
+    * @return the function
+    */
+  def apply(function: MiscellaneousFunctionId, funcParams: Seq[Any]): LabelTypeInstantFunction =
+  {
+    function match {
+      case LabelReplace       => LabelReplaceImpl(funcParams)
+      case _                  => throw new UnsupportedOperationException(s"$function not supported.")
+    }
+  }
+}
+
+case class LabelReplaceImpl(funcParams: Seq[Any]) extends LabelTypeInstantFunction {
```

Comment:
you could simply have this as a function within LabelFunctions object. Any specific need for separate class?

Reply:
Changed the design